### PR TITLE
Use visibility class for category product indexer, instead of hardcoded values

### LINKED ISSUE #37200: [Issue] Use visibility class for category product indexer, instead of hardcoded values
This issue is automatically created based on existing pull request: magento/magento2#37199: Use visibility class for category product indexer, instead of hardcoded values

---------

Replace hardcoded values by using proper class in category product indexer. This fix provides more modularity.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by Nuranto**
Replace hardcoded values by using proper class in category product indexer. This fix provides more modularity.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37200

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37200: Use visibility class for category product indexer, instead of hardcoded values

### CODE REVIEW

**Comment by Nuranto** [comment]
@magento create issue

**Comment by Nuranto** [comment]
@magento run all tests

**Comment by Nuranto** [comment]
@magento run Static Tests, Unit Tests, Integration Tests, Functional Tests EE, Functional Tests CE, Functional Tests B2B

File: app/code/Magento/Catalog/Model/Config.php [hunk 1]
```diff
@@ -114,6 +115,11 @@ class Config extends \Magento\Eav\Model\Config
      */
     protected $_configFactory;
 
+    /**
+     * @var Visibility
+     */
+    protected $_visibility;
```
Comment (Den4ik) [inline app/code/Magento/Catalog/Model/Config.php]: As I remember PSR don't recommend to use underscore in the property names

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @Nuranto Thanks for this code improvement. Please look at my comment.

File: app/code/Magento/Catalog/Model/Config.php (hunk 1 above)
Comment (Den4ik) [reply app/code/Magento/Catalog/Model/Config.php]: New properties should be private

File: app/code/Magento/Catalog/Model/Config.php [hunk 2]
```diff
@@ -147,6 +153,7 @@ public function __construct(
         \Magento\Eav\Model\ResourceModel\Entity\Attribute\Set\CollectionFactory $setCollectionFactory,
         \Magento\Store\Model\StoreManagerInterface $storeManager,
         \Magento\Eav\Model\Config $eavConfig,
+        Visibility $visibility,
```
Comment (Den4ik) [inline app/code/Magento/Catalog/Model/Config.php]: Please use [Backward Compatibility Documentation](https://developer.adobe.com/commerce/contributor/guides/code-contributions/backward-compatibility-policy/#adding-a-constructor-parameter) when add a new constructor params

File: app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php [hunk 3]
```diff
@@ -348,11 +348,7 @@ protected function getNonAnchorCategoriesSelect(Store $store)
                 \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED
             )->where(
                 $this->connection->getIfNullSql('cpvs.value', 'cpvd.value') . ' IN (?)',
-                [
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_IN_CATALOG,
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_IN_SEARCH,
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_BOTH
-                ]
+                $this->config->getVisibility()->getVisibleInSiteIds()
```
Comment (Den4ik) [inline app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php]: I believe that injecting `Visibility` model directly in this class is more correct solution follow [Backward Compatibility](https://developer.adobe.com/commerce/contributor/guides/code-contributions/backward-compatibility-policy/#adding-a-constructor-parameter)

Comment (Nuranto) [reply app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php]: Hi @Den4ik 

Thanks for your review, I applied your recommendations.

**Comment by Nuranto** [comment]
@magento run all tests

**Comment by Nuranto** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests, Unit Tests, WebAPI Tests

**Comment by Nuranto** [comment]
@magento run Functional Tests CE, Static Tests

**Comment by Nuranto** [comment]
@magento run all

**Comment by Nuranto** [comment]
@magento run all

**Comment by Nuranto** [comment]
@magento run all tests

**Comment by Nuranto** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests Ce, Functional Tests EE, Integration Tests, Sample Data Tests B2B, Static Tests, Unit Tests, WebAPI Tests

**Comment by Nuranto** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

**Comment by Nuranto** [comment]
@magento run Functional Tests B2B, Static Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Static Tests

**Review by engcom-Hotel — APPROVED** [review]
Hello @Nuranto,

Thanks for the contribution!

Changes are pretty straightforward and look good to me. Hence approving this PR for further process.

Failed tests seem flaky to me.

**Review by Den4ik — APPROVED** [review]
@engcom-Hotel Agree, all looks good

**Comment by engcom-November** [comment]
Hi @Nuranto,

Tried to update the pr to latest branch but got 403 error, could you please update the branch so we can move further with the pr.

**Comment by Nuranto** [comment]
done.

**Comment by engcom-November** [comment]
@magento run all tests

**Comment by engcom-November** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-November** [comment]
This is a code-quality pr, no manual testing is required.

As the tests are failing moving this to Extended testing.

**Comment by engcom-November** [comment]
The repeated failures for functional Tests B2B and EE are flaky. These are knows issues and are not related to PR changes.

**Functinal Tests B2B:**
**1st run:**
image

**2nd run:**
image

Known failure : [ACQE-6438](https://jira.corp.adobe.com/browse/ACQE-6438) - StorefrontPropertiesOfAProductAttributeTest

**Functinal Tests EE:**
**1st run:**
image

**2nd run:**
image

Known failure : [ACQE-6224](https://jira.corp.adobe.com/browse/ACQE-6224) - AdminContentHierarchyDisabledScopeWebsiteFilterTest

**Comment by engcom-November** [comment]
WebAPI Failures are not related to PR changes, these are known issues.

**WebAPI Tests:**
image

image

Known failure: [AC-11750](https://jira.corp.adobe.com/browse/AC-11750) - RetrieveOrdersByOrderNumberTest

Hence moving this to merge in progress.

**Comment by engcom-Charlie** [comment]
@engcom-November as per [this](https://github.com/magento/magento2/pull/37199#issuecomment-2068924186), I can see that there is one failure i.e StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest on which no data have been provided. Please provide more detail information on this. (mention open JIRA) till then moving it back to Extended Testing

**Comment by engcom-November** [comment]
@engcom-Charlie, I have updated the comment with the open jira of `StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest`, hence moving it back to Merge in Progress.

**Pull request merged**